tempesta_fw/vhost.c

```diff
@@ -247,11 +303,169 @@ tfw_vhost_match(TfwStr *arg)
 
 /*
  * Pointer to the current location structure.
- * The pointer is shared among several functions below.
+ * The pointer is shared among multiple functions below.
  */
 static TfwLocation *tfwcfg_this_location;
 
 /*
+ * Find a non-idempotent request definition entry within specified location.
+ * Entries are processed in the order they are defined in the configuration.
+ * That means the matching entry must be the last entry in the array, and it
+ * must have the same match @op and the same @arg.
+ */
+static TfwNipDef *
+tfw_nipdef_lookup(TfwLocation *loc, int op, char *arg, int len)
+{
+	TfwNipDef *nipdef;
+
+	if (!loc->nipdef_sz)
+		return NULL;
+
+	nipdef = loc->nipdef[loc->nipdef_sz - 1];
+	if ((nipdef->op == op) && (nipdef->len == len)
+	    && !strcasecmp(nipdef->arg, arg))
+		return nipdef;
+
+	return NULL;
+}
+
+static TfwNipDef *
+tfw_nipdef_lookup_dup(TfwLocation *loc, int method, int op, char *arg, int len)
+{
+	int i;
+	TfwNipDef *nipdef;
+
+	if (!loc->nipdef_sz)
+		return NULL;
+
+	/* Check all entries but the last one. */
+	for (i = 0; i < loc->nipdef_sz - 1; ++i) {
+		nipdef = loc->nipdef[i];
+		if ((nipdef->op == op) && (nipdef->len == len)
+		    && !strcasecmp(nipdef->arg, arg))
+			return nipdef;
+	}
+	/* Check the last entry. */
+	nipdef = loc->nipdef[i];
+	if ((nipdef->method & method) && (nipdef->op == op)
+	    && (nipdef->len == len) && !strcasecmp(nipdef->arg, arg))
+		return nipdef;
+
+	return NULL;
+}
+
+/*
+ * Create and initialize a new non-idempotent request definition entry,
+ * and add it to the given location structure. The entry is added as
+ * a pointer to the memory allocated to hold the definition.
+ */
+static TfwNipDef *
+tfw_nipdef_addnew(TfwLocation *loc, int method, int op, char *arg, int len)
+{
+	char *data;
+	TfwNipDef *nipdef;
+
+	if (loc->nipdef_sz == TFW_NIPDEF_ARRAY_SZ)
+		return NULL;
+
+	if ((data = kmalloc(sizeof(TfwNipDef) + len + 1, GFP_KERNEL)) == NULL)
+		return NULL;
+
+	nipdef = (TfwNipDef *)data;
+	nipdef->method = method;
+	nipdef->op = op;
+	nipdef->arg = data + sizeof(TfwNipDef);
+	nipdef->len = len;
+	memcpy((void *)nipdef->arg, (void *)arg, len + 1);
+
+	loc->nipdef[loc->nipdef_sz++] = nipdef;
+
+	return nipdef;
+}
+
+static int
+tfw_handle_nonidempotent(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int ret, len, method, op;
+	char *in_method, *in_op, *arg;
+	TfwLocation *loc = tfwcfg_this_location;
+	TfwNipDef *nipdef;
+
+	BUG_ON(!tfwcfg_this_location);
+
+	if (ce->attr_n) {
+		TFW_ERR("%s: Arguments may not have the \'=\' sign\n",
+			cs->name);
+		return -EINVAL;
+	}
+	if (ce->val_n != 3) {
+		TFW_ERR("%s: Invalid number of arguments.\n", cs->name);
+		return -EINVAL;
+	}
+
+	/* The method: one of GET, PUT, POST, etc. in form of a bitmask. */
+	in_method = (char *)ce->vals[0];
+	ret = tfw_cfg_map_enum(tfw_method_enum, in_method, &method);
```

Comment:
By the RFC 7231 4.2: 
```
Of the request methods defined by this specification, the GET, HEAD, OPTIONS, and 
TRACE methods are defined to be safe.
...
Of the request methods defined by this specification, PUT, DELETE, and safe request
methods are idempotent.
```

Maybe we should add new enum with methods that could be non-idempotent?